### PR TITLE
fixes #12621 - adding cache for template kinds to prevent multiple queries

### PR DESCRIPTION
**Pull request by unorthodoxgeek**

### CODE REVIEW

File: lib/template_kind_cache.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  class << self
+    def clear_cache!
+      @names = nil
+    end
+
+    def names
+      @names ||= TemplateKind.pluck(:name)
+    end
+  end
+end
```
Comment (shlomizadok) [inline lib/template_kind_cache.rb]: nitpick: end of line

File: app/models/template_kind.rb [hunk 2]
```diff
@@ -6,4 +6,12 @@ class TemplateKind < ActiveRecord::Base
   has_many :os_default_templates
   validates :name, :presence => true, :uniqueness => true
   scoped_search :on => :name
+
+  after_save :clear_cache!
```
Comment (shlomizadok) [inline app/models/template_kind.rb]: Not sure, but maybe add `if name.changed?` so it will clear cache on name changes?

Comment (unorthodoxgeek) [inline app/models/template_kind.rb]: don't think it's important enough a change to make, the less complex the better, IMHO

File: lib/template_kind_cache.rb (hunk 1 above)
Comment (unorthodoxgeek) [inline lib/template_kind_cache.rb]: added

**Comment by shlomizadok** [comment]
Works for me :+1:

**Comment by lzap** [comment]
Why don't you make model method `names_as_array` or `available_names` that caches the attribute? It does not look nice, but separate file for that looks like an overkill to me.

**Comment by unorthodoxgeek** [comment]
single responsibility is king, IMHO.
this isn't ActiveRecord territory, it's data caching, I think that this
solution is the best, but if you disagree I can change it.

**Comment by dLobatog** [comment]
@unorthodoxgeek Thanks for the concern about this problem. I wonder why Rails SQL caching isn't working here, queries like 'TemplateKind.all' ought to be cached. 

Why not just changing the line that calls `pluck` to this?

``` ruby
  Rails.cache.fetch("template_kinds", expires_in: 1.hour) do
      TemplateKind.pluck(:name)
 end
```

I didn't find a way to reproduce the bug though, how are you doing it?

**Comment by unorthodoxgeek** [comment]
@dLobatog - rails' SQL cache is a myth :smile: 
I've never seen it actually working.

Doing it with rails cache should do the trick, I prefer doing it using Ruby cache, but if you prefer it that way, I'm fine with changing it.

Regarding reproduction - remove apipie cache, and make a call to an apipie endpoint, see hundrends of queries.

**Comment by dLobatog** [comment]
@unorthodoxgeek I just prefer the Rails cache thing I wrote in the previous comment because it's simpler, fixes the problem and we don't have to test it. If you or someone has arguments for the 'DIY' cache that's fine too! I can't see any advantages for that approach here though :sweat_smile:

**Comment by unorthodoxgeek** [comment]
@dLobatog - okay. done.

**Comment by dLobatog** [comment]
Merged as 44ad385e0b64bd38feac0c1a0211e5ff7db05441, thanks @unorthodoxgeek!

**Pull request closed without merging**